test/test-ntt-avx512.cpp

```diff
@@ -232,6 +168,138 @@ TEST(NTT, WriteInvInterleavedT4) {
   AssertEqual(exp, out);
 }
 
+#ifdef HEXL_HAS_AVX512IFMA
+// First parameter is the NTT degree
+// Second parameter is the number of bits in the NTT modulus
+// Third parameter is whether or not to prefer small primes
+class DegreeModulusBoolTest
+    : public ::testing::TestWithParam<std::tuple<uint64_t, uint64_t, bool>> {
+ protected:
+  void SetUp() {}
+
+  void TearDown() {}
+
+ public:
+};
+
+TEST_P(DegreeModulusBoolTest, FwdNTTAVX512IFMA) {
+  if (!has_avx512ifma) {
+    GTEST_SKIP();
+  }
+  uint64_t N = std::get<0>(GetParam());
+  uint64_t modulus_bits = std::get<1>(GetParam());
+  bool prefer_small_primes = std::get<2>(GetParam());
+  uint64_t modulus = GeneratePrimes(1, modulus_bits, prefer_small_primes, N)[0];
+
+#ifdef HEXL_DEBUG
+  size_t num_trials = 1;
+#else
+  size_t num_trials = 20;
+#endif
+
+  std::random_device rd;
+  std::mt19937 gen(rd());
+  std::uniform_int_distribution<uint64_t> distrib(1, modulus - 1);
+
+  for (size_t trial = 0; trial < num_trials; ++trial) {
+    std::vector<uint64_t> input64(N, 0);
+    for (size_t i = 0; i < N; ++i) {
+      input64[i] = distrib(gen);
+    }
+
+    std::vector<uint64_t> input_ifma = input64;
+    std::vector<uint64_t> input_ifma_lazy = input64;
+    std::vector<uint64_t> exp_output(N, 0);
+
+    // Compute reference
+    NTT ntt64(N, modulus);
+    ReferenceForwardTransformToBitReverse(input64.data(), N, modulus,
+                                          ntt64.GetRootOfUnityPowers().data());
+
+    ForwardTransformToBitReverseAVX512<52>(
+        input_ifma.data(), N, ntt64.GetModulus(),
+        ntt64.GetAVX512RootOfUnityPowers().data(),
+        ntt64.GetAVX512Precon52RootOfUnityPowers().data(), 1, 1);
+
+    // Compute lazy
+    ForwardTransformToBitReverseAVX512<52>(
+        input_ifma_lazy.data(), N, ntt64.GetModulus(),
+        ntt64.GetAVX512RootOfUnityPowers().data(),
+        ntt64.GetAVX512Precon52RootOfUnityPowers().data(), 2, 4);
+    for (auto& elem : input_ifma_lazy) {
+      elem = elem % modulus;
+    }
+
+    AssertEqual(input64, input_ifma);
+    AssertEqual(input64, input_ifma_lazy);
+  }
+}
+
+TEST_P(DegreeModulusBoolTest, InvNTTAVX512IFMA) {
+  if (!has_avx512ifma) {
+    GTEST_SKIP();
+  }
+  uint64_t N = std::get<0>(GetParam());
+  uint64_t modulus_bits = std::get<1>(GetParam());
+  bool prefer_small_primes = std::get<2>(GetParam());
+  uint64_t modulus = GeneratePrimes(1, modulus_bits, prefer_small_primes, N)[0];
+
+#ifdef HEXL_DEBUG
+  size_t num_trials = 1;
+#else
+  size_t num_trials = 20;
+#endif
+
+  std::random_device rd;
+  std::mt19937 gen(rd());
+  std::uniform_int_distribution<uint64_t> distrib(1, modulus - 1);
+
+  for (size_t trial = 0; trial < num_trials; ++trial) {
+    std::vector<uint64_t> input64(N, 0);
+    for (size_t i = 0; i < N; ++i) {
+      input64[i] = distrib(gen);
+    }
```

Comment:
This code block is repeated a few times. Would it be worth, refactoring into a helper function that returns a vector? Or, abstract away distrib and gen into a functor and use std::generate.

Reply:
Thanks for the suggestion. I've created https://jiratest.idoc.intel.com/browse/GLADE-79 to fix in a separate PR